File: solutions/cpp/luhn/luhn.cpp
#include "luhn.h"

namespace luhn
{
    bool valid(const std::string &input)
    {
        int sum{0}, count{0}, factor{2};
        for (size_t i = input.size(); i > 0; i--)
        {
            char letter = input[i - 1];
            if (letter == ' ')
            {
                continue;
            }
            if (letter < '0' || letter > '9')
            {
                return false;
            }

            factor = 3 - factor;
            int digit = int(letter - '0') * factor;
            sum += (digit > 9) ? digit - 9 : digit;
            count += 1;
        }
        return count > 1 && sum % 10 == 0;
    }
} // namespace luhn

File: solutions/elixir/luhn/lib/luhn.ex
defmodule Luhn do
  @doc """
  Checks if the given number is valid via the luhn formula
  """
  @spec valid?(String.t()) :: boolean
  def valid?(number) do
    trimmed = number |> String.replace(~s/ /, "")

    trimmed =~ ~r/^\d{2,}$/ and
      [0 | get_digits(trimmed)]
      |> Enum.map_every(2, &double/1)
      |> Enum.sum()
      |> rem(10) == 0
  end

  @spec get_digits(String.t()) :: [integer]
  defp get_digits(number) do
    number
    |> String.graphemes()
    |> Enum.map(&String.to_integer/1)
    |> Enum.reverse()
  end

  @spec double(integer) :: integer
  defp double(n) when n < 5, do: n * 2
  defp double(n), do: n * 2 - 9
end
